src/coreclr/vm/ceeload.cpp

```diff
@@ -1078,17 +1073,13 @@ void Module::SetDebuggerInfoBits(DebuggerAssemblyControlFlags newBits)
     m_dwTransientFlags |= (newBits << DEBUGGER_INFO_SHIFT_PRIV);
 
 #ifdef DEBUGGING_SUPPORTED
-    BOOL setEnC = ((newBits & DACF_ENC_ENABLED) != 0) && IsEditAndContinueCapable();
-
-    // The only way can change Enc is through debugger override.
-    if (setEnC)
+    if (IsEditAndContinueCapable())
```

Comment:
The code depends on `IsEditAndContinueCapable` to be invariant, but it may actually change because of it depends on `GetDebuggerInfoBits` that the debugger can change. There is commented out assert in `BOOL IsEditAndContinueEnabled()` that suggest it may actually happen in some cases. I think that this can interact poorly with this change, depending on how the debugger decides to change the bits.

I think we should:
- Compute IsEditAndContinueCapable just once in `Module::Create`
- Use immutable cached value of IsEditAndContinueCapable everywhere else. It can be achieve by changing `BOOL IsEditAndContinueCapable()` to virtual method that returns false in `class Module` and returns true in `class EditAndContinueModule`.

Reply:
Also, enable the assert in `IsEditAndContinueEnabled` and change it to just `return (IsEditAndContinueCapable()) && ((m_dwTransientFlags & IS_EDIT_AND_CONTINUE) != 0);`

Reply:
Are we sure that after Module::Create the DACF_ALLOW_JIT_OPTS in the Assembly will not change? What about the debugger API (via  DacDbiInterfaceImpl::SetCompilerFlags)? Isn't the module already created by then? It is clear to me and it is confusing because Assembly, DomainAssembly and Module all have flags.

To double check what you are saying about IsEditAndContinueCapable():
```
Module:
  virtual BOOL IsEditAndContinueCapable() const { return FALSE; }
EditAndContinueModule:
  virtual BOOL IsEditAndContinueCapable() const { return TRUE; }
```
Module::Create() doesn't need any changes because it already creates an 
`EditAndContinueModule` instance if `IsEditAndContinueCapable(pAssembly, file)` returns true.

Reply:
I agree that it is confusing and hard to reason about.

> Module::Create() doesn't need any changes

Correct.